mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,746 @@
+      }
+
+
+      // Since the operations are dense, we can directly use BLAS calls here.
+      val deltaMean: BV[Double] = if (buffer.mean != null) {
+        require(other.mean != null)
+        b(other.mean) :- b(buffer.mean)
+      } else { null }
+
+      if (buffer.mean != null) {
+        require(other.mean != null)
+        require(weightSum != null)
+        b(buffer.mean) :+= deltaMean :* (b(weightSum2) :/ weightSum)
+      }
+
+      if (buffer.m2n != null) {
+        require(other.m2n != null)
+        val w = (b(weightSum1) :* b(weightSum2)) :/ weightSum
+        val z = (deltaMean :* deltaMean) :* w
+        b(buffer.m2n) :+= b(other.m2n) :+ z
+      }
+
+      if (buffer.m2 != null) {
+        require(other.m2 != null)
+        b(buffer.m2) :+= b(other.m2)
+      }
+
+      if (buffer.l1 != null) {
+        require(other.l1 != null)
+        b(buffer.l1) :+= b(other.l1)
+      }
+
+      if (buffer.max != null) {
+        require(other.max != null)
+        b(buffer.max) := la.max(b(buffer.max), b(other.max))
+      }
+
+      if (buffer.min != null) {
+        require(other.min != null)
+        b(buffer.min) := la.min(b(buffer.min), b(other.min))
+      }
+
+      if (buffer.nnz != null) {
+        require(other.nnz != null)
+        bl(buffer.nnz) :+= bl(other.nnz)
+      }
+
+      if (buffer.weightSum != null) {
+        require(other.weightSum != null)
+        b(buffer.weightSum) :+= b(other.weightSum)
+      }
+    }
+  }
+
+  private case class MetricsAggregate(
+      requested: Seq[Metrics],
+      startBuffer: Buffer,
+      child: Expression,
+      mutableAggBufferOffset: Int,
+      inputAggBufferOffset: Int)
+    extends TypedImperativeAggregate[Buffer] {
+
+    // These objects are not thread-safe, allocate them in the aggregator.
+    private[this] lazy val row = new UnsafeRow(Buffer.numFields)
+    private[this] lazy val projection = UnsafeProjection.create(Buffer.bufferSchema)
+
+
+    override def eval(buff: Buffer): InternalRow = {
+      val metrics = requested.map({
+        case Mean => UnsafeArrayData.fromPrimitiveArray(Buffer.mean(buff))
+        case Variance => UnsafeArrayData.fromPrimitiveArray(Buffer.variance(buff))
+        case Count => Buffer.totalCount(buff)
+        case NumNonZeros => UnsafeArrayData.fromPrimitiveArray(Buffer.nnz(buff))
+        case Max => UnsafeArrayData.fromPrimitiveArray(Buffer.max(buff))
+        case Min => UnsafeArrayData.fromPrimitiveArray(Buffer.min(buff))
+        case NormL2 => UnsafeArrayData.fromPrimitiveArray(Buffer.l2(buff))
+        case NormL1 => UnsafeArrayData.fromPrimitiveArray(Buffer.l1(buff))
+      })
+      InternalRow.apply(metrics: _*)
+    }
+
+    override def children: Seq[Expression] = child :: Nil
+
+    override def update(buff: Buffer, row: InternalRow): Buffer = {
+      // Unsafe rows do not play well with UDTs, it seems.
+      // Directly call the deserializer.
+      val v = udt.deserialize(row.getStruct(0, udt.sqlType.size))
```

Comment:
I think
```
val v = udt.deserialize(row.getStruct(0, udt.sqlType.size))
```
has some problems.
We cannot directly use `getter` method on `row` parameter passed in, because the ordinal of the column we want to get depends on the underlying catalyst layer, instead, we should use:
```
val datum = child.eval(row) 
val featureVector  = udt.deserialize(datum)
```
to get column value.